crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,723 @@
+use tedge_actors::LoggingReceiver;
+use tedge_actors::MessageBox;
+use tedge_actors::MessageReceiver;
+use tedge_actors::RuntimeError;
+use tedge_actors::RuntimeRequest;
+use tedge_actors::WrappedInput;
+use tedge_api::health::health_status_up_message;
+use tedge_api::OperationStatus;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_timer_ext::SetTimeout;
+use tedge_timer_ext::Timeout;
+use tedge_utils::file::move_file;
+use tedge_utils::file::PermissionEntry;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+use crate::config::FirmwareManagerConfig;
+use crate::error::FirmwareManagementError;
+use crate::message::get_child_id_from_child_topic;
+use crate::message::DownloadFirmwareStatusMessage;
+use crate::message::FirmwareOperationRequest;
+use crate::message::FirmwareOperationResponse;
+use crate::operation::ActiveOperationState;
+use crate::operation::FirmwareOperationEntry;
+use crate::operation::OperationKey;
+
+pub type OperationSetTimeout = SetTimeout<OperationKey>;
+pub type OperationTimeout = Timeout<OperationKey>;
+
+fan_in_message_type!(FirmwareInput[MqttMessage, OperationTimeout] : Debug);
+fan_in_message_type!(FirmwareOutput[MqttMessage, OperationSetTimeout] : Debug);
+
+pub struct FirmwareManagerActor {
+    config: FirmwareManagerConfig,
+    active_child_ops: HashMap<OperationKey, ActiveOperationState>,
+}
+
+#[async_trait]
+impl Actor for FirmwareManagerActor {
+    type MessageBox = FirmwareManagerMessageBox;
+
+    fn name(&self) -> &str {
+        "FirmwareManager"
+    }
+
+    async fn run(mut self, mut message_box: Self::MessageBox) -> Result<(), RuntimeError> {
+        self.resend_operations_to_child_device(&mut message_box)
+            .await?;
+        // TODO: Do we really need to send 500 from each actor?
+        self.get_pending_operations_from_cloud(&mut message_box)
+            .await?;
+
+        info!("Ready to serve the firmware request.");
+
+        while let Some(event) = message_box.recv().await {
+            match event {
+                FirmwareInput::MqttMessage(message) => {
+                    self.process_mqtt_message(message, &mut message_box).await?;
+                }
+                FirmwareInput::OperationTimeout(timeout) => {
+                    self.process_operation_timeout(timeout, &mut message_box)
+                        .await?;
+                } // TODO: Add downloader
+            }
+        }
+        Ok(())
+    }
+}
+
+impl FirmwareManagerActor {
+    pub fn new(config: FirmwareManagerConfig) -> Self {
+        Self {
+            config,
+            active_child_ops: HashMap::new(),
+        }
+    }
+
+    pub async fn process_mqtt_message(
+        &mut self,
+        message: MqttMessage,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        if self.config.health_check_topics.accept(&message) {
+            self.send_health_status_message(message_box).await?;
+            return Ok(());
```

Comment:
It is not required at the actor level, at least not yet. The health check contract is for a daemon and not for each and every actor. So, the main that's wrapping this actor will be the one that responds to the health-check requests, which is the c8y-device-management plugin in this case. When you refactor the existing c8y-firmware-plugin with this actor impl, you'll have another main that's expected to respond to tedge/health-check/c8y-firmware-plugin requests. Currently all such health-checks are handled by the tedge-health-check actor which is already wired with the c8y-device-management.

In future, we'd want the daemon/main to really health-check the underlying actors as well, instead of blindly sending a response immeidately.